pkg/hive/hive.go

```diff
@@ -218,3 +222,91 @@ func (s *Service) disconnect(peer p2p.Peer) error {
 
 	return nil
 }
+
+func (s *Service) startCheckPeersHandler() {
+	ctx, cancel := context.WithCancel(context.Background())
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		<-s.quit
+		cancel()
+	}()
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case newPeers := <-s.peersChan:
+				s.wg.Add(1)
+				go func() {
+					defer s.wg.Done()
+					s.checkAndAddPeers(ctx, newPeers)
+				}()
+			}
+		}
+	}()
+}
+
+func (s *Service) checkAndAddPeers(ctx context.Context, peers pb.Peers) {
+
+	var peersToAdd []swarm.Address
+	mtx := sync.Mutex{}
+	wg := sync.WaitGroup{}
+
+	for i := range peers.Peers {
+		err := s.sem.Acquire(ctx, 1)
+		if err != nil {
+			return
+		}
+
+		wg.Add(1)
+		go func(idx int) {
+			defer func() {
+				s.sem.Release(1)
+				wg.Done()
+			}()
+
+			newPeer := peers.Peers[idx]
```

Comment:
just a minor suggestion - since you are already iterating over `peers.Peers`, you could already:
```go
for _, p := range peers.Peers {
   go func(p bzz.Address){ ... } (p)
}
```